pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -33,10 +34,184 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
 )
 
 const resyncPeriod = 60 * time.Minute
 
+func updatePodNodePortCache(cache map[string][]k8s.PodNodePort, podIP string, nodePort, podPort int) {
+	v := k8s.PodNodePort{
+		NodePort: nodePort,
+		PodPort:  podPort,
+	}
+	_, exists := cache[podIP]
+	if !exists {
+		cache[podIP] = []k8s.PodNodePort{v}
+		return
+	}
+	cache[podIP] = append(cache[podIP], v)
+}
+
+// rulesToPodPortMap is responsible for going through the NPL rules and figures out if some of these
+// rules are still needed. To check if a rule is needed, it first checks if the relevant Pod exists,
+// if not, it deletes the rule. This function adds the port rules to the portTable and responds with
+// a list of visited Pods.
+func rulesToPodPortMap(portTable *portcache.PortTable, kubeClient clientset.Interface) (map[string][]k8s.PodNodePort, error) {
+	podPortRules, err := portTable.PodPortRules.GetAllRules()
+	if err != nil {
+		klog.Errorf("error in fetching the Pod port rules: %s", err.Error())
+		return nil, errors.New("error in fetching the Pod port rules: " + err.Error())
+	}
+
+	podNodePortCache := make(map[string][]k8s.PodNodePort)
+
+	for k, v := range podPortRules {
+		// k: NodePort
+		// v: {PodIP,Pod port}
+		if k > portTable.EndPort || k < portTable.StartPort {
+			// delete this rule, as the Node port falls in an invalid range
+			err := portTable.PodPortRules.DeleteRule(k, v.PodIP)
+			if err != nil {
+				klog.Errorf("error in deleting the rule for port %d and Pod IP %s", k, v.PodIP)
+				continue
+			}
+		}
+		podList, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
+			FieldSelector: "status.podIP=" + v.PodIP,
+		})
+		if err != nil {
+			klog.Infof("error in fetching Pod with IP address %s: %s", v.PodIP, err.Error())
+			continue
+		}
+		if len(podList.Items) == 0 {
+			klog.Infof("couldn't find a Pod for IP address %s, will delete it's entry", v.PodIP)
+			// delete this port's entry from the rules
+			err := portTable.PodPortRules.DeleteRule(k, v.PodIP)
+			if err != nil {
+				// TODO: need to handle this case properly, currently, it would mean that Antrea thinks
+				// this port is free to allocate, however the rule for this port forwards the packets
+				// to a non-existent Pod
+				klog.Errorf("error in deleting the rule for port %d and Pod IP %s: %s", k, v.PodIP, err.Error())
+			}
+			continue
+		}
+		// Pod exists, add it to podPortTable
+		// error is already checked
+		klog.V(2).Infof("adding an entry for Node port %d, Pod %s and Pod port: %d", k, v.PodIP, v.PodPort)
+		portTable.AddUpdateEntry(k, v.PodPort, v.PodIP)
```

Comment:
Do we need to check if the Pod has NPL enabled, or you will handle this by the controller later?